### PR TITLE
Recipify: RoR Final Project Version

### PR DESCRIPTION
**Pull request by jicamargo**
# **Recipify: RoR Final Project Version**

## **Description:**
This pull request represents the final version of the Recipify project, incorporating the required features and additional functionalities. The project is designed to manage recipes, food lists for users.

**Authors:** @gdumani, @jicamargo 

## **Project Requirements:**
1. **Login and Registration Pages:**
   - Implemented using Devise for user authentication.

2. **Food List:**
   - Displays a list of food items added by the logged-in user.
   - Quantity of each food item is displayed.
   - Allows users to add new food items.

3. **Recipes List:**
   - Displays a list of recipes created by the logged-in user.
   - Leads to recipe details.
   - Enables recipe deletion for the recipe owner.

4. **Public Recipe List:**
   - Displays a list of all public recipes, ordered by newest.
   - Leads to recipe details.

5. **Recipe Details:**
   - Allows toggling a recipe between public and private status.
   - Displays recipe details if public or if the user is the owner.
   - Leads to a form for adding new food items if the user is the owner.

6. **General Shopping List View:**
   - Shows the list of missing food items required for all recipes of the logged-in user.
   - Counts total food items and total price of missing items.
   - Ensures no N+1 queries are present.

7. **Navigation Menu:**
   - Provides a navigation menu to access all project pages.

**Technical Setup:**
- GitHub repository set up with Gitflow.
- Devise integrated for user authentication.
- RSpec and Capybara testing libraries configured.
- Bullet check for N+1, eager load Issues

This pull request represents the culmination of our efforts in developing Recipify, ensuring it meets the specified project requirements and provides additional features for enhanced functionality.

Please review the code changes, features, and confirm the successful implementation of the project requirements.

Thank you for your time and consideration.

**Reviewers:** @gdumani, @jicamargo

### CODE REVIEW

File: app/controllers/foods_controller.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  # DELETE /foods/1 or /foods/1.json
+  def destroy
+    @food.destroy
+
+    respond_to do |format|
+      format.html { redirect_to foods_url, notice: 'Food was successfully deleted' }
+      format.json { head :no_content }
+    end
+  end
```
Comment (VICTOR-CHUKWU) [inline app/controllers/foods_controller.rb:61]: - [x] When food is added to the shopping list and I try to delete that food, I have an error which is not user friendly, kindly fix this
Screenshot from 2023-09-15 22-08-21

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  # DELETE /recipes/1 or /recipes/1.json
+  def destroy
+    @recipe.destroy
+
+    respond_to do |format|
+      format.html { redirect_to recipes_url, notice: 'Recipe was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
```
Comment (VICTOR-CHUKWU) [inline app/controllers/recipes_controller.rb:71]: - [x] When I add a recipe to my shopping list and try to delete it, I get an error, kindly fix that
Screenshot from 2023-09-15 22-08-34

File: spec/requests/public_recipes_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+RSpec.describe 'Recipes Index Page', type: :system do
+  add_info_to_db
+
+  scenario 'User can see the list of recipes' do
+    login(user.email, user.password)
+    visit '/recipes_public'
+    sleep(1) # Wait for the page to load
+
+    expect(page).to have_content('Public Recipes')
+    expect(page).to have_content('Test Recipe')
```
Comment (VICTOR-CHUKWU) [inline spec/requests/public_recipes_spec.rb:13]: - [x] a test is failing, kindly fix it
image

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @gdumani  and @jicamargo  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good and Fantastic design and styling
- Good commit history
- Clean and DRY codebase


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/foods_controller.rb (hunk 1 above)
Comment (jicamargo) [reply app/controllers/foods_controller.rb:61]: Done! Thank you

File: app/controllers/recipes_controller.rb (hunk 2 above)
Comment (jicamargo) [reply app/controllers/recipes_controller.rb:71]: Done! Thank you

File: spec/requests/public_recipes_spec.rb (hunk 3 above)
Comment (jicamargo) [reply spec/requests/public_recipes_spec.rb:13]: Done! Thank you

**Review by DuaneDave — APPROVED** [review]
Hi @jicamargo and @gdumani 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✔️ Beautiful design
✔️ I like the feedback the app gives a user per action

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**